src/services.rs

```diff
@@ -0,0 +1,93 @@
+use std::sync::{Arc, Mutex};
+use std::collections::HashMap;
+use multimap::MultiMap;
+use rand::{Rng, thread_rng};
+use dns_parser::{self, QueryClass, Name, RRData};
+
+pub type AnswerBuilder = dns_parser::Builder<dns_parser::Answers>;
+
+
+pub type SharedServices = Arc<Mutex<Services>>;
+
+pub struct Services {
+    pub hostname: Name<'static>,
+    /// main index
+    pub by_id: HashMap<usize, ServiceData>,
+    /// maps to id
+    pub by_type: MultiMap<Name<'static>, usize>,
+    /// maps to id
+    pub by_name: HashMap<Name<'static>, usize>
+}
+
+impl Services {
```

Comment:
This could have a `find_by_name` / `find_by_type`, which return a `Box<Iterator<&ServiceData>>`, to be used by `handle_question`.

The hashmaps can probably be made private after that.